atlasdb-cassandra/src/test/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainerTest.java

```diff
@@ -0,0 +1,325 @@
+import static org.mockito.Mockito.when;
+
+import com.palantir.atlasdb.keyvalue.cassandra.ReloadingCloseableContainer;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import com.palantir.refreshable.Refreshable;
+import com.palantir.refreshable.SettableRefreshable;
+import java.io.Closeable;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+import org.awaitility.Awaitility;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ReloadingCloseableContainerTest {
+    private static final ExecutorService executor =
+            PTExecutors.newCachedThreadPool(ReloadingCloseableContainerTest.class.getName());
+
+    private static final int INITIAL_VALUE = 0;
+    private static final int UPDATED_VALUE = 1;
+
+    @Mock
+    private Closeable mockCloseable;
+
+    @Mock
+    private Closeable mockCloseableExtra;
+
+    @Mock
+    private Function<Integer, Closeable> closeableFactory;
+
+    private SettableRefreshable<Integer> settableRefreshable;
+
+    @Before
+    public void setUp() {
+        settableRefreshable = Refreshable.create(INITIAL_VALUE);
+        when(closeableFactory.apply(INITIAL_VALUE)).thenReturn(mockCloseable);
+        when(closeableFactory.apply(UPDATED_VALUE)).thenReturn(mockCloseableExtra);
+    }
+
+    @Test
+    public void closeBlocksDuringRunWithInvocation() throws BrokenBarrierException, InterruptedException {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+        CyclicBarrier runWithFutureRunsFirstBarrier = new CyclicBarrier(2);
+        CyclicBarrier manuallyFinishRunWith = new CyclicBarrier(2);
+
+        executor.execute(() -> container.executeWithResource(closeable -> {
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            awaitOrFail(manuallyFinishRunWith);
+            verifyNoInteractions(closeable);
+        }));
+
+        // This test could pass spuriously, if close is just not called.
+        // The likelihood if this is minimised by
+        // 1. waiting on a barrier for the executeWithResource and,
+        // 2. Requiring a whole second of close not completing.
+        Future<?> closeFuture = executor.submit(() -> {
+            awaitOrFail(runWithFutureRunsFirstBarrier);
+            container.close();
+        });
+        Awaitility.await().during(Duration.ofSeconds(1)).until(() -> !closeFuture.isDone());
```

Comment:
this is largely pointless - the future could complete nearly immediately (which would be a bug), but the awaitility could still finish because the future initially is _not_ done. This doesn't actually test a value that could usefully change over time: either the future is done immediately, and this fails, or it is not done immediately, in which case this passes. This would make more sense if you _wanted_ the future to complete within a second (as you know that it would eventually go from not done to done at _some point in time_).

Reply:
For posterity and context: [`await().during()`](https://github.com/awaitility/awaitility/wiki/Usage#assert-that-a-value-is-maintained) asserts that a value is _maintained_ for a full second.

I acknowledge that this test can definitely pass spuriously (mentioned in a comment above the highlighted block of code), which is definitely not ideal and appreciate the solutions to fix that! 

In this particular case, the future completing nearly immediately would still fail. The future would need to not complete for a full second. Now, if the thread was just blocked arbitrarily (e.g not scheduled, rather than blocked on close) for more than a full second this would pass.

Reply:
Fair enough, I've never actually seen `await().during()`, but I think for good reason: it makes your test take _at least_ that long. I think we should redesign it such that we never need to do these kinds of waits (in that we _may_ have to wait some time for a thread to run a task, but not that we _must_ wait some period of time).